distbench_test_sequencer_test.cc

```diff
@@ -1095,6 +1095,99 @@ TEST(DistBenchTestSequencer, PreCheckInvalidActivityConfig) {
+
+  TestSequenceParams params;
+  params.nb_cliques = nb_cliques;
+
+  TestSequence test_sequence;
+  auto* test = test_sequence.add_tests();
+
+  auto* client = test->add_services();
+  client->set_name("client");
+  client->set_count(1);
+  client->set_protocol_driver_options_name("lo_opts");
+
+  auto* server = test->add_services();
+  server->set_name("server");
+  server->set_count(1);
+  server->set_protocol_driver_options_name("lo_opts");
+
+  auto* rpc_desc = test->add_rpc_descriptions();
+  rpc_desc->set_name("client_server_rpc");
+  rpc_desc->set_client("client");
+  rpc_desc->set_server("server");
+  rpc_desc->set_distribution_config_name("MyPayloadDistribution");
+
+  auto* client_al = test->add_action_lists();
+  client_al->set_name("client");
+  client_al->add_action_names("run_queries");
+
+  auto* server_al = test->add_action_lists();
+  server_al->set_name("client_server_rpc");
+
+  auto* lo_opts = test->add_protocol_driver_options();
+  lo_opts->set_name("lo_opts");
+  lo_opts->set_netdev_name("lo");
+
+  auto* req_dist = test->add_distribution_configs();
+  req_dist->set_name("MyPayloadDistribution");
+  for (float i = 1; i < 5; i++) {
+    auto* pmf_point = req_dist->add_pmf_points();
+    pmf_point->set_pmf(i / 10);
+
+    auto* data_point = pmf_point->add_data_points();
+    data_point->set_exact(i * 11);
+
+    data_point = pmf_point->add_data_points();
+    data_point->set_exact(i * 9);
+  }
+  req_dist->add_field_names("request_payload_size");
+  req_dist->add_field_names("response_payload_size");
+
+  auto action = test->add_actions();
+  action->set_name("run_queries");
+  action->set_rpc_name("client_server_rpc");
+  action->mutable_iterations()->set_max_iteration_count(20);
+
+  // TODO: Delete this message in final version.
+  LOG(ERROR) << test_sequence.DebugString();
+
+  TestSequenceResults results;
+  auto context = CreateContextWithDeadline(/*max_time_s=*/75);
+  grpc::Status status = tester.test_sequencer_stub->RunTestSequence(
+      context.get(), test_sequence, &results);
+  ASSERT_OK(status);
+
+  // TODO: Delete this message in final version.
+  LOG(ERROR) << "Test Results Are: "
+             << results.test_results(0).service_logs().DebugString();
+
+  ASSERT_EQ(results.test_results().size(), 1);
+  auto& test_results = results.test_results(0);
+  ASSERT_EQ(test_results.service_logs().instance_logs_size(), 1);
+  const auto& instance_results_it =
+      test_results.service_logs().instance_logs().find("client/0");
+  ASSERT_NE(instance_results_it,
+            test_results.service_logs().instance_logs().end());
+
+  int num_samples = 0;
+  for (const auto& rpc_sample : instance_results_it->second.peer_logs()
+                                    .find("server/0")
+                                    ->second.rpc_logs()
+                                    .find(0)
+                                    ->second.successful_rpc_samples()) {
+    ASSERT_EQ(rpc_sample.request_size() % 11, 0);
```

Comment:
So this test will succeed if the RNG always returns 0. We should probably make the pass/fail more realistic.

Reply:
That will not happen because it is immediately followed by     
`ASSERT_NE(rpc_sample.request_size(), 0);`

Reply:
I see what the problem is - The change I mentioned above was added in [this commit ](https://github.com/google/distbench/pull/156/commits/c8a8684bb3387eceab9d1cb53f36dd8e4a5c1385#diff-edec1e9a4d7bd7befdc78ca86a613bec65bb0aac87dfb0f6a74bc0cdfefb5894)

Reply:
So there is nothing that checks that the distribution is actually returning a range of values, only that it is returning an allowed value. If it returns the same value over and over the test passes.

Reply:
Returning same value over and over is a valid scenario where this value got really lucky.

Whether a distribution returns correct values or not should be outside the scope of a distbench_test_sequencer_test.

This is actually checked for in tests inside the randomization directory.

Reply:
Can you point me at this other unit test?

Reply:
Check out the test - `FullTestPmf` in `randomization/distribution_sample_generator_test.cc`

Github is not letting me get a link to specific link of the file.

Reply:
https://github.com/google/distbench/blob/Add_Randomization_Lib_To_Distbench/randomization/distribution_sample_generator_test.cc#L174

This is testing a < b < c. It would be a better test it it tested |a-E(a)| < delta, |b-E(b)| < delta, |c-E(c)| < delta

Reply:
Hmmm, I will give this a try.

Reply:
Added this.

https://github.com/google/distbench/pull/156/commits/91746596663ba5c345f7319ff7f165a3b6f3edc6